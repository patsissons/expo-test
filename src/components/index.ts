export * from './state';
export * from './actions';
export * from './context';
export * from './AppComponent';
export * from './startup/Startup';
