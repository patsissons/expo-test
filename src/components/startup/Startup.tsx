import React from 'react';
import { View, ActivityIndicator } from 'react-native';
import { Store, AppComponent, withStore } from '..';

export namespace Startup {
  export interface Props {
    header?: any;
    activityIndicator?: any;
  }

  export interface ActivityIndicatorComponentProps {
    animating?: boolean;
    color?: string;
    hidesWhenStopped?: boolean;
    size?: number | 'small' | 'large';
  }

  export interface ComponentProps extends AppComponent.Props<Store>, ActivityIndicatorComponentProps, Props {
  }
}

class StartupComponent extends AppComponent<Startup.ComponentProps, Store> {
  static defaultProps: Partial<Startup.ComponentProps> = {
    size: 'large',
  };

  componentDidMount() {
    this.props.actions.context.loadEnv();
    this.props.actions.context.loadLocale();
  }

  render() {
    if (this.isLoading) {
      return (
        <View style={ this.S.common.container }>
          { this.R.coerceToText(this.props.header) }
          { this.renderActivityIndicator() }
        </View>
      );
    }

    return this.props.children;
  }

  protected renderActivityIndicator() {
    if (this.props.activityIndicator) {
      return this.props.activityIndicator;
    }

    const { animating, color, hidesWhenStopped, size } = this.props;

    return (
      <ActivityIndicator { ...{ animating, color, hidesWhenStopped, size } } />
    );
  }

  protected get isLoading() {
    return [
      this.props.context == null,
      this.props.context.env == null,
      this.props.context.locale == null,
      this.props.context.localization == null,
    ].some(x => x);
  }
}

export const Startup = withStore(
  StartupComponent,
);
