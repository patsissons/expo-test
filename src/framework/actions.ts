import { Action } from 'redux';
import { NavigationParams } from 'react-navigation';
import { createTypeNameMap, isString } from '../utils';

export type Actions = Readonly<typeof actions>;

export interface RequestAction<T = {}> extends Action {
  payload?: T;
}

export interface LocalePayload {
  locale: string;
}

export interface DemoPayload {
  amount: number;
}

export function isLocalePayload(payload: any): payload is LocalePayload {
  const localePayload: LocalePayload = payload;

  return (
    localePayload != null &&
    isString(localePayload.locale)
  );
}

export function getRequestAction(type: string): RequestAction;
export function getRequestAction<T>(type: string, payload: T): RequestAction<T>;
export function getRequestAction<T>(type: string, payload?: T): RequestAction<T> {
  return {
    type,
    payload,
  };
}

enum Context {
  LoadEnv,
  LoadLocale,
  LoadLocalization,
}

enum Navigation {
  NavigationBack,
  NavigationToMain,
  NavigationToLogin,
}

enum Demo {
  DemoIncrement,
  DemoDecrement,
}

export const actionTypes = {
  context: createTypeNameMap(Context),
  navigation: createTypeNameMap(Navigation),
  demo: createTypeNameMap(Demo),
};

export const actions = {
  context: {
    loadEnv: () => getRequestAction(actionTypes.context.LoadEnv),
    loadLocale: () => getRequestAction(actionTypes.context.LoadLocale),
  },
  navigation: {
    goBack: (params?: NavigationParams) => getRequestAction(actionTypes.navigation.NavigationBack, params),
    toMain: (params?: NavigationParams) => getRequestAction(actionTypes.navigation.NavigationToMain, params),
    toLogin: (params?: NavigationParams) => getRequestAction(actionTypes.navigation.NavigationToLogin, params),
  },
  demo: {
    increment: (payload?: DemoPayload) => getRequestAction(actionTypes.demo.DemoIncrement, payload),
    decrement: (payload?: DemoPayload) => getRequestAction(actionTypes.demo.DemoDecrement, payload),
  },
};
