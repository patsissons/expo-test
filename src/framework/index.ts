export * from './state';
export * from './actions';
export * from './context';
export * from './styles';
export * from './renderHelpers';
export * from './AppComponent';
