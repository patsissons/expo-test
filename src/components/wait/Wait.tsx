import React from 'react';
import { View, ActivityIndicator } from 'react-native';
import { Store, AppComponent, withStore } from '../../framework';

export namespace Wait {
  export interface Props {
    isWaiting: (store: Store) => boolean;
    init?: Function;
    header?: any;
    activityIndicator?: any;
  }

  export interface ActivityIndicatorComponentProps {
    animating?: boolean;
    color?: string;
    hidesWhenStopped?: boolean;
    size?: number | 'small' | 'large';
  }

  export interface ComponentProps extends AppComponent.Props<Store>, ActivityIndicatorComponentProps, Props {
  }
}

class WaitComponent extends AppComponent<Wait.ComponentProps, Store> {
  static defaultProps: Partial<Wait.ComponentProps> = {
    size: 'large',
  };

  componentDidMount() {
    if (this.props.init) {
      this.props.init();
    }
  }

  render() {
    if (this.props.isWaiting(this.props.state)) {
      return (
        <View style={ this.S.common.container }>
          { this.R.coerceToText(this.props.header) }
          { this.renderActivityIndicator() }
        </View>
      );
    }

    return this.props.children;
  }

  protected renderActivityIndicator() {
    if (this.props.activityIndicator) {
      return this.props.activityIndicator;
    }

    const { animating, color, hidesWhenStopped, size } = this.props;

    return (
      <ActivityIndicator { ...{ animating, color, hidesWhenStopped, size } } />
    );
  }
}

export const Wait = withStore(
  WaitComponent,
);
