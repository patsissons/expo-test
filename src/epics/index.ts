import 'rxjs';
import { Action } from 'redux';
import { combineEpics, Epic } from 'redux-observable';
import { Store } from '../framework';
import { loadLocale } from './context';
import { navigate } from './navigation';

export const epics: Epic<Action, Store> = combineEpics(
  loadLocale,
  navigate,
);
