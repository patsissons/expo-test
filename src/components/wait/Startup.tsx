import React from 'react';
import { Store, AppComponent, withStore } from '../../framework';
import { Wait } from './Wait';

export namespace Startup {
  export interface Props {
  }

  export interface ComponentProps extends AppComponent.Props<Store>, Props {
  }
}

class StartupComponent extends AppComponent<Startup.ComponentProps, Store> {
  render() {
    return (
      <Wait init={ this.init.bind(this) } isWaiting={ this.isWaiting.bind(this) }>
        { this.props.children }
      </Wait>
    );
  }

  protected init() {
    this.props.actions.context.loadEnv();
    this.props.actions.context.loadLocale();
  }

  protected isWaiting(store: Store) {
    return [
      store.context == null,
      store.context.env == null,
      store.context.locale == null,
      store.context.localization == null,
    ].some(x => x);
  }
}

export const Startup = withStore(
  StartupComponent,
);
