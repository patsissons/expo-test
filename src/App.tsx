import React from 'react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import { createEpicMiddleware } from 'redux-observable';
import { composeWithDevTools } from 'redux-devtools-extension'
import '../types';
import { reducers } from './reducers';
import { epics } from './epics';
import { AppNavigator, reactNavigationMiddleware } from './navigation/AppNavigator';

const store = createStore(
  reducers,
  composeWithDevTools(
    applyMiddleware(
      createEpicMiddleware(epics as any),
      reactNavigationMiddleware,
    ),
  ),
);

export class App extends React.Component {
  render() {
    return (
      <Provider store={ store }>
        <AppNavigator />
      </Provider>
    );
  }
}
