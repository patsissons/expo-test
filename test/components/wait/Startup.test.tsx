import React from 'react';
import { shallow } from 'enzyme';
import { createStoreProps } from '../../storeUtils';
import { actionTypes } from '../../../src/framework';
import { Startup } from '../../../src/components/wait/Startup';

describe('components', () => {
  describe('Startup', () => {
    it('renders loading mode without crashing', () => {
      // Arrange
      const storeProps = createStoreProps({ context: {} });
      const component = (<Startup { ...storeProps as any } />);

      // Act
      const result = shallow(component).dive();

      // Assert
      expect(result).toMatchSnapshot('loading');
    });

    it('renders loaded mode without crashing', () => {
      // Arrange
      const loadedState = {
        context: {
          env: {},
          locale: 'test',
          localization: {},
        },
      };
      const storeProps = createStoreProps(loadedState);
      const component = (<Startup { ...storeProps as any } />);

      // Act
      const result = shallow(component).dive();

      // Assert
      expect(result).toMatchSnapshot('loading');
    });

    it('loads context on init', () => {
      // Arrange
      const storeProps = createStoreProps({});
      const component = (<Startup { ...storeProps as any } />);

      // Act
      const result = shallow(component).dive();
      const instance: any = result.instance();
      instance.init();

      // Assert
      expect(storeProps.store.getActions()).toContainEqual({ type: actionTypes.context.LoadEnv });
      expect(storeProps.store.getActions()).toContainEqual({ type: actionTypes.context.LoadLocale });
    });

    it('waits while context is not fully loaded', () => {
      // Arrange
      const storeProps = createStoreProps({});
      const store = storeProps.store.getState();
      const component = (<Startup { ...storeProps as any } />);

      // Act
      const result = shallow(component).dive();
      const instance: any = result.instance();
      store.context = {} as any;
      const unloaded = instance.isWaiting(store);
      store.context = {
        env: {},
        locale: 'test',
        localization: {},
      } as any;
      const loaded = instance.isWaiting(store);

      // Assert
      expect(unloaded).toBeTruthy();
      expect(loaded).toBeFalsy();
    });
  });
});
