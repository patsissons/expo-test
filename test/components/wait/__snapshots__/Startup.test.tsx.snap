// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Startup renders loaded mode without crashing: loading 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StartupComponent
    actions={
      Object {
        "context": Object {
          "loadEnv": [Function],
          "loadLocale": [Function],
        },
        "demo": Object {
          "decrement": [Function],
          "increment": [Function],
        },
        "navigation": Object {
          "goBack": [Function],
          "toLogin": [Function],
          "toMain": [Function],
        },
      }
    }
    context={
      Object {
        "env": Object {},
        "locale": "test",
        "localization": Object {},
      }
    }
    state={
      Object {
        "context": Object {
          "env": Object {},
          "locale": "test",
          "localization": Object {},
        },
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": undefined,
      "init": [Function],
      "isWaiting": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": undefined,
        "init": [Function],
        "isWaiting": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`components Startup renders loading mode without crashing: loading 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StartupComponent
    actions={
      Object {
        "context": Object {
          "loadEnv": [Function],
          "loadLocale": [Function],
        },
        "demo": Object {
          "decrement": [Function],
          "increment": [Function],
        },
        "navigation": Object {
          "goBack": [Function],
          "toLogin": [Function],
          "toMain": [Function],
        },
      }
    }
    context={Object {}}
    state={
      Object {
        "context": Object {},
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": undefined,
      "init": [Function],
      "isWaiting": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": undefined,
        "init": [Function],
        "isWaiting": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
